File: src/Python/replace_matrix.py
###########################################################
# @Author: HoodUSSEnterprise
# @Date: 2026-06-22 13:08:22
# @LastEditors: HoodUSSEnterprise
# @LastEditTime: 2026-06-22 13:16:20
# @FilePath: \asm_matrix_benchmark\src\Python\replace_matrix.py
# @Description:replace matrix python code
###########################################################

from typing import Union

from .base_matrix import Matrix
from .find_matrix import find_elem


###########################################################
# @description:  replace matrix by coord
# @return {*}
###########################################################
def replace_pos(
    m: "Matrix", x_label: int, y_label: int, new_value: Union[int, float]
) -> None:
    if x_label < 0 or x_label > m.cols - 1 or y_label < 0 or y_label > m.rows - 1:
        print(f"Invalid coord ({x_label}, {y_label})")
        return

    m.data[x_label * m.cols + y_label] = new_value


###########################################################
# @description: replace matrix by element
# @return {*}
###########################################################
def replace_elem(
    m: "Matrix", old_value: Union[int, float], new_value: Union[int, float]
) -> None:
    if find_elem(m, old_value) != (-1, -1):
        print(f"No find number {old_value}")
        return
    while find_elem(m, old_value) != (-1, -1):
        x, y = find_elem(m, old_value)
        m.data[x * m.cols + y] = new_value

File: src/C++/replace_matrix.cpp
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-23 17:39:18
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-23 19:44:26
@FilePath: \asm_matrix_benchmark\src\C++\replace_matrix.cpp
@Description: replace matrix c++ code
*************************************************************/
#include "matrix_cpp.h"
#include <stdexcept>

/***********************************************************
@description:Replaces the element at the specified (x, y) position with a new value
@param {size_t} x
@param {size_t} y
@param {int} new_value
@return {*}
*************************************************************/
void MatrixInt::replace_pos(size_t x, size_t y, int new_value)
{
    if (x >= rows || y >= cols)
    {
        throw std::invalid_argument("replace_pos: coordinate out of range");
    }
    data[x * cols + y] = new_value;
}

/***********************************************************
@description:Replaces all occurrences of a specific value with a new value throughout the entire matrix
@param {int} old_value
@param {int} new_value
@return {*}
*************************************************************/
void MatrixInt::replace_elem(int old_value, int new_value)
{
    for (size_t i = 0; i < rows * cols; i++)
    {
        if (data[i] == old_value)
        {
            data[i] = new_value;
        }
    }
}
